### DISCUSSION TITLE
How would I go about implementing `ValidBigDecimal` for `String`?

### DISCUSSION THREAD

**Discussion by KaranAhlawat** [Iltotore/iron#191] [Q&A] [answered]
Or is there something of the sort in the library already? I mostly want to translate `String Refined ValidBigDecimal` from scala-refined to iron.

**Comment by KaranAhlawat** [comment]
Update: I tried writing a custom constraint, but that doesn't work because try-catch cannot be inlined. Neither can the `scala.util.Try` data structure. Any ideas? @Iltotore

**Reply by Iltotore** [reply, marked as answer]
If you want to make your constraint usable at compile-time, you will probably have to use a macro.

You can look into `io.github.iltotore.iron.constraint.string` for examples.

PS: thank you for the bump, I forgot about your thread 😅

**Reply by KaranAhlawat** [reply]
Thank you! I think I that was it. Here is what I came up with (and it seems to work)

```scala
  final class ValidBigDecimal

  inline given Constraint[String, ValidBigDecimal] with
    override inline def message: String = "Should be a valid decimal"

    override inline def test(value: String): Boolean =
      ${check('value)}

  private def check(expr: Expr[String])(using Quotes): Expr[Boolean] =
    expr.value match
      case Some(value) =>
        Expr(Try(BigDecimal(value)).isSuccess)
      case _ =>
        '{Try(BigDecimal($expr)).isSuccess}
```

**Reply by Iltotore** [reply]
Looks good to me :+1:
